File: java/std_nat/src/fan/std/ThreadLocal.java
package fan.std;

import fan.sys.FanObj;
import fan.sys.Func;
import fanx.main.Sys;
import fanx.main.Type;

public class ThreadLocal extends FanObj {
	private final java.lang.ThreadLocal local;

	private static Type type = null;

	public Type typeof() {
		if (type == null) {
			type = Sys.findType("std::ThreadLocal");
		}
		return type;
	}

	public ThreadLocal(final Func init) {
		if (init == null) {
			local = new java.lang.ThreadLocal();
		} else {
			local = new java.lang.ThreadLocal() {
				@Override
				protected Object initialValue() {
					return init.call();
				}
			};
		}
	}

	public static ThreadLocal make(Func initail) {
		ThreadLocal self = new ThreadLocal(initail);
		return self;
	}

	public Object get() {
		return local.get();
	}

	public ThreadLocal set(Object val) {
		local.set(val);
		return this;
	}

	public void remove() {
		local.remove();
	}
}

File: library/std/js/ThreadLocal.js


fan.std.ThreadLocal = fan.sys.Obj.$extend(fan.sys.Obj);
fan.std.ThreadLocal.prototype.$ctor = function() {}
fan.std.ThreadLocal.prototype.$typeof = function() { return fan.std.ThreadLocal.$type; }

fan.std.ThreadLocal.make = function(init) {
	var self = new fan.std.ThreadLocal();
	if (init != null) self.m_val = init.call();
	else self.m_val = null;
	return self;
}

fan.std.ThreadLocal.prototype.get = function() {
	return this.m_val;
}

fan.std.ThreadLocal.prototype.set = function(val) {
	this.m_val = val;
}

fan.std.ThreadLocal.prototype.remove = function() {
	this.m_val = null;
}


